modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessExecuteAction.java

```diff
@@ -0,0 +1,341 @@
+            }
+            builder.endObject();
+            return builder;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            Request request = (Request) o;
+            return Objects.equals(script, request.script) &&
+                context == request.context;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(script, context);
+        }
+
+        public enum SupportedContext {
+
+            EXECUTE_API_SCRIPT((byte) 0);
+
+            private final byte id;
+
+            SupportedContext(byte id) {
+                this.id = id;
+            }
+
+            public static SupportedContext fromId(byte id) {
+                switch (id) {
+                    case 0:
+                        return EXECUTE_API_SCRIPT;
+                    default:
+                        throw new IllegalArgumentException("unknown context [" + id + "]");
+                }
+            }
+        }
+
+    }
+
+    public static class RequestBuilder extends ActionRequestBuilder<Request, Response, RequestBuilder> {
+
+        RequestBuilder(ElasticsearchClient client) {
+            super(client, INSTANCE, new Request());
+        }
+    }
+
+    public static class Response extends ActionResponse implements ToXContentObject {
+
+        private Object result;
+
+        Response() {}
+
+        Response(Object result) {
+            this.result = result;
+        }
+
+        public Object getResult() {
+            return result;
+        }
+
+        @Override
+        public void readFrom(StreamInput in) throws IOException {
+            super.readFrom(in);
+            result = in.readGenericValue();
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            out.writeGenericValue(result);
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            builder.field("result", result);
+            return builder.endObject();
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            Response response = (Response) o;
+            return Objects.equals(result, response.result);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(result);
+        }
+    }
+
+    // This was first an interface, but changed it to abstract class in order to add the params field / getter and constructor.
+    // I wonder why I need to do this? The fact that Factory#newInstance(...) accepts params is already explicit enough?
+    // This just feels to me as unneeded boilerplate code.
```

Comment:
I assume this comment was added to remember the question and should be removed before committing.  @rjernst can likely explain the necessity of the design decisions here better than I can, so I'm hopeful he will chime in :)

Reply:
The factory is what holds onto the params, so that they can be passed to the constructor. Think of the factory as the signature for the constructor. It's not boilerplate; it is actually needed based on current uses of scripts throughout the system. Also note that the factory signature is what allows the script instance to have arbitrary objects passed in. If `ScriptService.compile` were to return an instance directly, instead of a factory, we would need some way to pass in this information in a generic way, which would probably mean duck typing through a String->Object map and then require casts. With the factory, we get static type checking of the arguments a script needs to be constructed.

Reply:
Thanks for the explanation.

I think I added this after looking at `ExecutableScript`, but I guess there `setNextVar()` method is used to provide `params` to the script engine (so `params`  isn't implicitly provided).